crates/cairo-annotations/src/annotations/profiler.rs

```diff
@@ -29,12 +32,84 @@ pub struct ProfilerAnnotationsV1 {
+    static ref RE_MONOMORPHIZATION: Regex = Regex::new(r"<.*>")
+        .expect("Failed to create regex normalising mononorphised generic functions names");
+}
+
 /// The fully qualified Cairo path of the Cairo function.
 #[derive(
     Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Deserialize, Display, Default,
 )]
 pub struct FunctionName(pub String);
 
+impl FunctionName {
+    #[must_use]
+    pub fn from_sierra_statement_idx(
+        statement_idx: StatementIdx,
+        sierra_program: &Program,
+        split_generics: bool,
+    ) -> Self {
+        // The `-1` here can't cause an underflow as the statement id of first function's entrypoint is
+        // always 0, so it is always on the left side of the partition, thus the partition index is > 0.
+        let function_idx = sierra_program
+            .funcs
+            .partition_point(|f| f.entry_point.0 <= statement_idx.0)
+            - 1;
+        let function_name = sierra_program.funcs[function_idx].id.to_string();
+        // Remove suffix in case of loop function e.g. `[expr36]`.
+        let function_name = RE_LOOP_FUNC.replace(&function_name, "");
+        // Remove parameters from monomorphised Cairo generics e.g. `<felt252>`.
+        let function_name = if split_generics {
+            function_name
+        } else {
+            RE_MONOMORPHIZATION.replace(&function_name, "")
+        };
+
+        Self(function_name.to_string())
+    }
+
+    /// `contract_name` and `function_name` are always present (in case they are not in trace we just
+    /// set `<unknown>` string).
+    /// `address` and `selector` are optional and set if `--show-details` flag is enabled
+    /// or names are unknown.
+    #[must_use]
+    pub fn from_entry_point_params(
+        contract_name: Option<String>,
+        function_name: Option<String>,
+        contract_address: ContractAddress,
+        function_selector: EntryPointSelector,
+        show_details: bool,
+    ) -> Self {
+        let (contract_name, address) = match contract_name {
+            Some(name) if show_details => (name, Some(contract_address.0)),
+            Some(name) => (name, None),
+            None => (String::from("<unknown>"), Some(contract_address.0)),
+        };
+
+        let (function_name, selector) = match function_name {
+            Some(name) if show_details => (name, Some(function_selector.0)),
+            Some(name) => (name, None),
+            None => (String::from("<unknown>"), Some(function_selector.0)),
+        };
+
+        let contract_address = match address {
+            None => String::new(),
+            Some(address) => format!("Address: {address}\n"),
+        };
+        let selector = match selector {
+            None => String::new(),
+            Some(selector) => format!("Selector: {selector}\n"),
+        };
+
+        FunctionName(format!(
+            "Contract: {contract_name}\n{contract_address}Function: {function_name}\n{selector}",
```

Comment:
I am not sure about this because this function doesn't return a valid function name. Maybe we should leave this in the profiler. What do you think?

Reply:
What do you mean by "valid function name"?

Reply:
The documentation for `FunctionName` states: `The fully qualified Cairo path of the Cairo function.` This would be something like `snforge_std::_cheatcode::_is_config_run` or `scarb_template::fib`. However, in this function, we use a string that doesn't uphold this contract, and you end up with function names in different formats.

Reply:
I would go the other way around. Instead of creating a type that holds two types of functions, I would create two separate types. This could even be a struct with fields for `contract`, `function`, and `selector`.

Reply:
discussed offline, agreed to move back to profiler since it;s not that important